mithril-client-cli/src/commands/cardano_stake_distribution/download.rs

```diff
@@ -0,0 +1,289 @@
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "Can not fetch Cardano stake distribution from unique identifier: '{}'",
+                    &self.unique_identifier
+                )
+            })?;
+
+        progress_printer.report_step(
+            2,
+            "Fetching the certificate and verifying the certificate chain…",
+        )?;
+        let certificate = client
+            .certificate()
+            .verify_chain(&cardano_stake_distribution.certificate_hash)
+            .await
+            .with_context(|| {
+                format!(
+                    "Can not verify the certificate chain from certificate_hash: '{}'",
+                    &cardano_stake_distribution.certificate_hash
+                )
+            })?;
+
+        progress_printer.report_step(
+            3,
+            "Verify that the Cardano stake distribution is signed in the associated certificate",
+        )?;
+        let message = MessageBuilder::new()
+            .compute_cardano_stake_distribution_message(&certificate, &cardano_stake_distribution)
+            .with_context(|| {
+                "Can not compute the message for the given Cardano stake distribution"
+            })?;
+
+        if !certificate.match_message(&message) {
+            return Err(anyhow!(
+                    "Certificate and message did not match:\ncertificate_message: '{}'\n computed_message: '{}'",
+                    certificate.signed_message,
+                    message.compute_hash()
+                ));
+        }
+
+        progress_printer.report_step(4, "Writing fetched Cardano stake distribution to a file")?;
+        if !download_dir.is_dir() {
+            std::fs::create_dir_all(download_dir)?;
+        }
+        let filepath = PathBuf::new().join(download_dir).join(format!(
+            "cardano_stake_distribution-{}.json",
+            cardano_stake_distribution.epoch
+        ));
+        std::fs::write(
+            &filepath,
+            serde_json::to_string(&cardano_stake_distribution).with_context(|| {
+                format!(
+                    "Can not serialize Cardano stake distribution artifact '{:?}'",
+                    cardano_stake_distribution
+                )
+            })?,
+        )?;
+
+        if self.json {
+            println!(
+                r#"{{"cardano_stake_distribution_epoch": "{}", "filepath": "{}"}}"#,
+                cardano_stake_distribution.epoch,
+                filepath.display()
+            );
+        } else {
+            println!(
+                "Cardano stake distribution for epoch '{}' has been verified and saved as '{}'.",
+                cardano_stake_distribution.epoch,
+                filepath.display()
+            );
+        }
+
+        Ok(())
+    }
+
+    fn is_sha256_hash(identifier: &str) -> bool {
+        identifier.len() == 64 && identifier.chars().all(|c| c.is_ascii_hexdigit())
```

Comment:
You should tests the property of the hash described here (len and hexdigit) :).

Reply:
Thanks, done in this separate commit: https://github.com/input-output-hk/mithril/pull/1883/commits/9f916381f91a37e8ff49e296b7359bf3847e593e